chips/apollo3/src/stimer.rs

```diff
@@ -181,14 +181,18 @@ impl<'a> Alarm<'a> for STimer<'a> {
         regs.stcfg
             .modify(STCFG::COMPARE_A_EN::CLEAR + STCFG::COMPARE_B_EN::CLEAR);
 
-        if !now.within_range(reference, expire) {
-            expire = now;
-        }
-
         // Enable interrupts
         regs.stminten
             .modify(STMINT::COMPAREA::SET + STMINT::COMPAREB::SET);
 
+        if !now.within_range(reference, expire) {
+            // The alarm has already expired!
+            // Let's set the interrupt manually
+            regs.stcfg.modify(STCFG::COMPARE_A_EN::SET);
+            regs.stmintset.modify(STMINT::COMPAREA::SET);
+            return;
+        }
+
```

Comment:
This can race with retrieving now and setting the alarm. We should conservatively estimate the number of clock cycles that can pass between querying `now` on line 172 and setting it here, and add these cycles to the `now` we're checking against.

Reply:
It can't really though as the CPU runs a lot faster then the timer clock. I added a comment to clarify this

Reply:
You could retrieve `now` right before the timer clock ticks, and then do the register write right after. That'll wait for a full timer wraparound, won't it?

Reply:
We already handle that with the

```rust
        if timer_delta < self.minimum_dt() {
            timer_delta = self.minimum_dt();
            expire = now.wrapping_add(timer_delta);
        }
```

Although obviously that should also handle an expiry in the past, so the `minimum_dt()` is too small I think

Reply:
I bumped `minimum_dt()` so that there shouldn't be any race conditions if `now` does tick

I kept the  "in the past check" (the main change here) so if the timer has already occurred we just generate the interrupt instead of mucking around setting an alarm as it's a bit clunky with all of the erratas

Reply:
@lschuermann ping

Reply:
Sorry, it took me some time to convince myself that this code is correct now. Alarm's a really hard system to think through all edge cases.